aws-android-sdk-appsync/src/main/java/com/amazonaws/mobileconnectors/appsync/AWSAppSyncDeltaSync.java

```diff
@@ -0,0 +1,601 @@
+                boolean executeBaseQuery = true;
+                if ( !forceFetch && deltaQuery != null) {
+                    //Check if
+                    long deltaInSeconds = (System.currentTimeMillis() - (lastRunTimeInMilliSeconds - 2000)) / 1000;
+                    Log.v(TAG, "Delta Sync: Time since last sync [" + deltaInSeconds + "] seconds");
+                    if (deltaInSeconds < baseRefreshIntervalInSeconds) {
+                        Log.v(TAG, "The last baseQuery from NETWORK was executed less than [" + baseRefreshIntervalInSeconds + "] seconds ago. Will run DeltaQuery from network");
+                        executeBaseQuery = false;
+                    }
+                    else {
+                        Log.v(TAG, "The last baseQuery from NETWORK run was before [" + baseRefreshIntervalInSeconds + "] seconds. Will run BaseQuery from network");
+                    }
+                }
+                else {
+                    Log.v(TAG, "Will run BaseQuery from network");
+                }
+
+                if (executeBaseQuery) {
+                    runBaseQuery(AppSyncResponseFetchers.NETWORK_ONLY);
+                }
+                else {
+                    runDeltaQuery();
+                }
+
+                if (deltaSyncOperationFailed) {
+                    scheduleRetry();
+                    return;
+                }
+
+                //Propagate all messages received so far and put subscription in processing mode
+                synchronized (processingLock) {
+                    Log.v(TAG, "Delta Sync: Delta query completed. Will propagate any queued messages on subscription");
+                    Response r;
+                    while ((r = messageQueue.poll()) != null) {
+                        if (subscriptionCallback != null ) {
+                            Log.v(TAG, "Delta Sync: Propagating queued message on subscription");
+                            subscriptionCallback.onResponse(r);
+                        }
+                    }
+                    Log.d(TAG, "Delta Sync: All queued messages propagated. Flipping mode to PROCESSING");
+                    mode = PROCESSING_MODE;
+                }
+
+                retryAttempt = 0;
+            }
+        }).start();
+
+        return id;
+    }
+
+    void cancel() {
+         Log.i(TAG, "Delta Sync: Cancelling Delta Sync operation [" + id + "]");
+        cancelled = true;
+        if (deltaSyncSubscriptionWatcher != null ) {
+            deltaSyncSubscriptionWatcher.cancel();
+        }
+        if (nextRetryAttempt != null ) {
+            nextRetryAttempt.cancel(true);
+            nextRetryAttempt = null;
+        }
+        if (nextScheduledRun != null ) {
+            nextScheduledRun.cancel(true);
+            nextScheduledRun = null;
+        }
+        deltaSyncObjects.remove(id);
+    }
+
+    public static void cancel(Long id) {
```

Comment:
Should this method be protected/ internal?

Reply:
Good call. It is in a class that is package private, but wouldn't hurt to make the method package private as well. Will fix.